File: OLSource1/v1/VS_officedev/codesnippet/VisualBasic/persisting-dynamic-controls-in-office-documents_3.vb
    Private Sub Application_DocumentBeforeClose(ByVal Doc As Word.Document, _
        ByRef Cancel As Boolean) Handles Application.DocumentBeforeClose

        Dim isExtended As Boolean = Globals.Factory.HasVstoObject(Doc)

        If isExtended Then

            Dim vstoDocument As Document = Globals.Factory.GetVstoObject(Doc)

            Dim controlsToRemove As System.Collections.ArrayList = _
                New System.Collections.ArrayList()

            ' Get all of the Windows Forms controls.
            For Each control As Object In vstoDocument.Controls
                If TypeOf control Is System.Windows.Forms.Control Then
                    controlsToRemove.Add(control)
                End If
            Next

            ' Remove all of the Windows Forms controls from the document.
            For Each control As Object In controlsToRemove
                vstoDocument.Controls.Remove(control)
            Next
        End If
    End Sub

File: OLSource1/v1/VS_officedev/codesnippet/CSharp/persisting-dynamic-controls-in-office-documents_3.cs
        void Application_DocumentBeforeClose(Word.Document Doc, ref bool Cancel)
        {

            bool isExtended = Globals.Factory.HasVstoObject(Doc);

            if (isExtended)
            {
                Microsoft.Office.Tools.Word.Document vstoDocument = Globals.Factory.GetVstoObject(Doc);

                
                System.Collections.ArrayList controlsToRemove = 
                    new System.Collections.ArrayList();
                
                // Get all of the Windows Forms controls.
                foreach (object control in vstoDocument.Controls)
                {
                    if (control is System.Windows.Forms.Control)
                    {
                        controlsToRemove.Add(control);
                    }
                }

                // Remove all of the Windows Forms controls from the document.
                foreach (object control in controlsToRemove)
                {
                    vstoDocument.Controls.Remove(control);
                }
            }
        }